x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java

```diff
@@ -183,13 +189,24 @@ public void document(int docID, StoredFieldVisitor visitor) throws IOException {
         };
     }
 
+    static Map<String, Object> filter(Map<String, ?> map, CharacterRunAutomaton includeAutomaton) {
+        return filter(map, includeAutomaton, 0);
+    }
+
     /** Filter a map by a {@link CharacterRunAutomaton} that defines the fields to retain. */
     @SuppressWarnings("unchecked")
-    static Map<String, Object> filter(Map<String, ?> map, CharacterRunAutomaton includeAutomaton, int initialState) {
+    private static Map<String, Object> filter(Map<String, ?> map, CharacterRunAutomaton includeAutomaton, int initialState) {
         Map<String, Object> filtered = new HashMap<>();
         for (Map.Entry<String, ?> entry : map.entrySet()) {
             String key = entry.getKey();
 
+            // only do this for `initialState == 0` since metadata fields would always only be top-level fields
```

Comment:
noob question: Wondering if there are any guarantees for this? Is this enforced somewhere or we just know that whitelisted ones are always top-level?

Reply:
good q! we just know, based on compiling the allowlist -- it's semi-enforced in code now in the sense that `static Map<String, Object> filter(Map<String, ?> map, CharacterRunAutomaton includeAutomaton)` is the only package-private method which always passes 0 for initial state.